File: src/engine/normalization.ts
import { RawEvent, NormalizedEvent } from '../types.js';

export function normalizeEmail(email: any): string {
  if (!email || typeof email !== 'string') return '';
  const trimmed = email.trim().toLowerCase();
  // Basic sanity check
  if (!trimmed.includes('@') || !trimmed.includes('.')) {
    return '';
  }
  return trimmed;
}

export function normalizePhone(phone: any): string {
  if (!phone || typeof phone !== 'string') return '';
  const trimmed = phone.trim();
  // Keep '+' if it starts with one, then extract all digits
  const hasLeadingPlus = trimmed.startsWith('+');
  const digitsOnly = trimmed.replace(/\D/g, '');
  if (!digitsOnly) return '';
  return hasLeadingPlus ? `+${digitsOnly}` : digitsOnly;
}

export function normalizeName(name: any): string {
  if (!name || typeof name !== 'string') return '';
  // Trim and collapse multiple spaces into one
  const collapsed = name.trim().replace(/\s+/g, ' ');
  if (!collapsed) return '';

  // Title-case words nicely while preserving clean presentation
  return collapsed
    .split(' ')
    .map(word => {
      if (word.length === 0) return '';
      return word.charAt(0).toUpperCase() + word.slice(1).toLowerCase();
    })
    .join(' ');
}

export function normalizeSource(source: any): string {
  if (!source || typeof source !== 'string') return 'website';
  return source.trim().toLowerCase();
}

export function normalizeEventType(eventType: any): string {
  if (!eventType || typeof eventType !== 'string') return 'page_visit';
  return eventType.trim().toLowerCase();
}

export function normalizeCampaign(campaign: any): string {
  if (!campaign || typeof campaign !== 'string') return '';
  // Normalize whitespace and trim
  return campaign.trim();
}

export function normalizeTimestamp(timestamp: any): string {
  if (!timestamp || typeof timestamp !== 'string') return new Date().toISOString();
  try {
    const d = new Date(timestamp.trim());
    return d.toISOString();
  } catch {
    return String(timestamp).trim();
  }
}

export function normalizePayload(payload: any): Record<string, any> {
  if (!payload) return {};
  if (typeof payload === 'object' && !Array.isArray(payload)) {
    return payload;
  }
  if (typeof payload === 'string') {
    try {
      const parsed = JSON.parse(payload);
      return typeof parsed === 'object' && parsed !== null ? parsed : { raw: payload };
    } catch {
      return { raw: payload };
    }
  }
  return { raw: payload };
}

export function normalizeEvent(raw: RawEvent, arrivalIndex: number = 0): NormalizedEvent {
  const event_id = String(raw.event_id || '').trim();
  const email = normalizeEmail(raw.email);
  const phone = normalizePhone(raw.phone);
  const name = normalizeName(raw.name);
  const source = normalizeSource(raw.source);
  const campaign = normalizeCampaign(raw.campaign);
  const event_type = normalizeEventType(raw.event_type);
  const timestamp = normalizeTimestamp(raw.timestamp);
  const status = raw.status ? String(raw.status).trim() : '';
  const lead_id = raw.lead_id ? String(raw.lead_id).trim() : '';
  const payload = normalizePayload(raw.payload);

  return {
    event_id,
    lead_id,
    email,
    phone,
    name,
    source,
    campaign,
    event_type,
    timestamp,
    status,
    payload,
    arrival_index: arrivalIndex,
  };
}

export function normalizeEventBatch(rawEvents: RawEvent[]): NormalizedEvent[] {
  return rawEvents.map((raw, idx) => normalizeEvent(raw, idx));
}

File: src_py/normalization.py
import re
import json

def normalize_email(email) -> str:
    if not email or not isinstance(email, str):
        return ""
    trimmed = email.strip().lower()
    if "@" not in trimmed or "." not in trimmed:
        return ""
    return trimmed

def normalize_phone(phone) -> str:
    if not phone or not isinstance(phone, str):
        return ""
    trimmed = phone.strip()
    has_plus = trimmed.startswith('+')
    digits = re.sub(r'\D', '', trimmed)
    if not digits:
        return ""
    return f"+{digits}" if has_plus else digits

def normalize_name(name) -> str:
    if not name or not isinstance(name, str):
        return ""
    collapsed = re.sub(r'\s+', ' ', name.strip())
    if not collapsed:
        return ""
    words = collapsed.split(' ')
    return ' '.join(w.capitalize() for w in words if w)

def normalize_source(source) -> str:
    if not source or not isinstance(source, str):
        return "website"
    return source.strip().lower()

def normalize_event_type(event_type) -> str:
    if not event_type or not isinstance(event_type, str):
        return "page_visit"
    return event_type.strip().lower()

def normalize_campaign(campaign) -> str:
    if not campaign or not isinstance(campaign, str):
        return ""
    return campaign.strip()

def normalize_payload(payload) -> dict:
    if not payload:
        return {}
    if isinstance(payload, dict):
        return payload
    if isinstance(payload, str):
        try:
            parsed = json.loads(payload)
            return parsed if isinstance(parsed, dict) else {"raw": payload}
        except Exception:
            return {"raw": payload}
    return {"raw": str(payload)}

def normalize_event(raw: dict, arrival_index: int = 0) -> dict:
    return {
        "event_id": str(raw.get("event_id") or "").strip(),
        "lead_id": str(raw.get("lead_id") or "").strip(),
        "email": normalize_email(raw.get("email")),
        "phone": normalize_phone(raw.get("phone")),
        "name": normalize_name(raw.get("name")),
        "source": normalize_source(raw.get("source")),
        "campaign": normalize_campaign(raw.get("campaign")),
        "event_type": normalize_event_type(raw.get("event_type")),
        "timestamp": str(raw.get("timestamp") or "").strip(),
        "status": str(raw.get("status") or "").strip(),
        "payload": normalize_payload(raw.get("payload")),
        "arrival_index": arrival_index
    }
